src/mono/browser/runtime/http.ts

```diff
@@ -66,167 +81,188 @@ export function http_wasm_abort_response(res: ResponseExtension): void {
-        await Promise.race([ts.__writer.ready, ts.__fetch_promise]);
-        await Promise.race([ts.__writer.close(), ts.__fetch_promise]);
+        await Promise.race([controller.streamWriter.ready, controller.fetchResponsePromise]);
+        await Promise.race([controller.streamWriter.close(), controller.fetchResponsePromise]);
     });
 }
 
-export function http_wasm_transform_stream_abort(ts: TransformStreamExtension): void {
-    ts.__writer.abort();
-}
-
-export function http_wasm_fetch_stream(url: string, header_names: string[], header_values: string[], option_names: string[], option_values: any[], abort_controller: AbortController, body: TransformStreamExtension): ControllablePromise<ResponseExtension> {
-    const fetch_promise = http_wasm_fetch(url, header_names, header_values, option_names, option_values, abort_controller, body.readable);
-    body.__fetch_promise = fetch_promise;
+export function http_wasm_fetch_stream(controller: HttpController, url: string, header_names: string[], header_values: string[], option_names: string[], option_values: any[]): ControllablePromise<void> {
+    if (BuildConfiguration === "Debug") commonAsserts(controller);
+    const transformStream = new TransformStream<Uint8Array, Uint8Array>();
+    controller.streamWriter = transformStream.writable.getWriter();
+    const fetch_promise = http_wasm_fetch(controller, url, header_names, header_values, option_names, option_values, transformStream.readable);
     return fetch_promise;
 }
 
-export function http_wasm_fetch_bytes(url: string, header_names: string[], header_values: string[], option_names: string[], option_values: any[], abort_controller: AbortController, bodyPtr: VoidPtr, bodyLength: number): ControllablePromise<ResponseExtension> {
+export function http_wasm_fetch_bytes(controller: HttpController, url: string, header_names: string[], header_values: string[], option_names: string[], option_values: any[], bodyPtr: VoidPtr, bodyLength: number): ControllablePromise<void> {
+    if (BuildConfiguration === "Debug") commonAsserts(controller);
     // the bodyPtr is pinned by the caller
     const view = new Span(bodyPtr, bodyLength, MemoryViewType.Byte);
     const copy = view.slice() as Uint8Array;
-    return http_wasm_fetch(url, header_names, header_values, option_names, option_values, abort_controller, copy);
+    return http_wasm_fetch(controller, url, header_names, header_values, option_names, option_values, copy);
 }
 
-export function http_wasm_fetch(url: string, header_names: string[], header_values: string[], option_names: string[], option_values: any[], abort_controller: AbortController, body: Uint8Array | ReadableStream | null): ControllablePromise<ResponseExtension> {
+export function http_wasm_fetch(controller: HttpController, url: string, header_names: string[], header_values: string[], option_names: string[], option_values: any[], body: Uint8Array | ReadableStream | null): ControllablePromise<void> {
```

Comment:
Could the return type still be `ControllablePromise<Response>`? Then you don't need the cast at the return statement.

Reply:
I will do it the other way around, because on the C# side, we don't read the resolved value.